lib/datadog/profiling/collectors/cpu_and_wall_time_worker.rb

```diff
@@ -62,7 +62,6 @@ def stop(*_)
 
             return unless @worker_thread
 
-            @worker_thread.kill
```

Comment:
So we're not actively killing `@worker_thread` anymore, merely relying on the `join` below to wait for it to disappear?

What is stopping `@worker_thread` then? Is that part of a side effect of `self.class._native_stop(self)`?

Reply:
The worker thread is "stuck" inside `_native_sampling_loop`, which **releases the Global VM Lock** and calls out to `run_sampling_trigger_loop` which is doing

```c
  while (state->should_run) {
    state->stats.trigger_sample_attempts++;

    // TODO: This is still a placeholder for a more complex mechanism. In particular:
    // * We want to signal a particular thread or threads, not the process in general
    // * We want to track if a signal landed on the thread holding the global VM lock and do something about it
    // * We want to do more than having a fixed sampling rate

    kill(getpid(), SIGPROF);
    nanosleep(&time_between_signals, NULL);
  }
```

The `_native_stop` sets `state->should_run = false` and so makes it come out of this loop, and thus return and finish cleanly.

Calling `@worker_thread.kill` effectively did the same -- because the worker thread runs that loop without the global VM lock, and we give the VM an "interruption function", which it calls when it wants to force us to stop what we're doing:

```c
// This is called by the Ruby VM when it wants to shut down the background thread
static void interrupt_sampling_trigger_loop(void *state_ptr) {
  struct cpu_and_wall_time_worker_state *state = (struct cpu_and_wall_time_worker_state *) state_ptr;

  state->should_run = false;
}
```

The difference between `_native_stop` and `Thread#kill` is only what happens when the thread re-acquires the global VM lock before coming back to Ruby land.

With `_native_stop`, the thread acquires the GVL normally and returns from `_native_sampling_loop` back to the Ruby code you see in `cpu_and_wall_time_worker.rb`. With `Thread#kill`, the VM will behave as if an exception was raised, and `_native_sampling_loop` will never return back to Ruby code.